### PR TITLE
Add GDExtension C++ snippets to tutorials/best_practices

### PR DESCRIPTION
**Pull request by dementive**
C++ documentation for tutorials/best_practices. I mostly have this done but it definitely needs some feedback/review/testing.

General question about the rest of the cpp documentation going forward: Should I add the GDCLASS, empty constructor, and _bind_methods to every class?

These are technically required for the C++ code to compile. However, adding these makes it harder for the reader to get the point the documentation is actually trying to make because they have to parse through a lot more boilerplate code. So im not sure it is actually useful to include in snippets outside of the initial tutorial documentation.

# Progress

1. scene_organization.rst - Done

2. scenes_versus_scripts.rst - Decided writing these snippets in C++ isn't really relevant. GDExtension doesn't really have a concept of "scripts" since everything is an actual Node/Object.

3. godot_interfaces.rst - TODO

4. godot_notifications.rst - Done, some of the code in the first block isn't correct. Might want to check how I did the `_unhandled_input` to make sure it is ideal, this is how I've been doing it in my code but not sure if it is necessarily the "correct" way to do it even though it does work.

5. data_preferences.rst - This isn't really entirely relevant to C++ either. Should probably be a separate GDExtension article for this that explains all the data structures in godot-cpp/templates and this article should link to that new article. Since a majority of the time (at least in my code) you are using the template data structures instead of the Variant one's unless you are exporting the code to gdscript.

6. logic_preferences.rst - Done

7. autoloads_versus_internal_nodes.rst - There should probably be a note that GDExtension actually does use singletons via `Engine::get_singleton()->register_singleton`. Maybe a new GDExtension example page on how to create singletons because its pretty common.

I can get the new singleton and data structures articles started if we think they would be good to add, but might want to do it in a separate PR idk what would be best.

### CODE REVIEW

File: tutorials/best_practices/godot_notifications.rst [hunk 1]
```diff
@@ -102,6 +102,27 @@ implementing a Timer-timeout loop is another option.
...
+    public:
+        // Allows for recurring operations that don't trigger script logic
+        // every frame (or even every fixed frame).
+        void _ready() override {
+            Timer *timer = memnew(Timer());
+            timer->set_autostart(true);
+            timer->set_wait_time(0.5);
+            add_child(timer);
+
+            // I have no idea what this would be in C++  :(
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:119]: You should add a method directly, as we don't have a callable for lambdas in c++

File: tutorials/best_practices/godot_notifications.rst [hunk 2]
```diff
@@ -160,6 +181,27 @@ delta time methods as needed.
...
+    class MyNode : public Node {
+    public:
+        // Called every frame, even when the engine detects no input.
+        void _process(double p_delta) {
+            if (Input::get_singleton->is_action_just_pressed("ui_select"))
+                UtilityFunctions::print(p_delta);
+        }
+
+        // Called during every input event. Equally true for _input().
+        void _unhandled_input(const Ref<InputEvent> &event) {
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:197]:
```suggestion
        void _unhandled_input(const Ref<InputEvent> &p_event) {
```
Prefix arguments with `p_` in c++on

File: tutorials/best_practices/godot_notifications.rst [hunk 3]
```diff
@@ -160,6 +181,27 @@ delta time methods as needed.
...
 
+  .. code-tab:: cpp C++
+
+    using namespace godot;
+
+    class MyNode : public Node {
+    public:
+        // Called every frame, even when the engine detects no input.
+        void _process(double p_delta) {
+            if (Input::get_singleton->is_action_just_pressed("ui_select"))
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:192]:
```suggestion
            if (Input::get_singleton->is_action_just_pressed("ui_select")) {
```
Always use brackets

File: tutorials/best_practices/godot_notifications.rst [hunk 4]
```diff
@@ -160,6 +181,27 @@ delta time methods as needed.
...
+                UtilityFunctions::print(p_delta);
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:193]:
```suggestion
                UtilityFunctions::print(p_delta);
            }
```

File: tutorials/best_practices/godot_notifications.rst [hunk 5]
```diff
@@ -309,3 +379,36 @@ nodes that one might create at runtime.
             GD.Print("I'm reacting to my parent's interaction!");
         }
     }
+
+  .. code-tab:: cpp C++
+
+    using namespace godot;
+
+    class MyNode : public Node {
+      Node parent_cache;
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:388]: Should use four spaces for indentation, throughout this file

File: tutorials/best_practices/scene_organization.rst [hunk 6]
```diff
@@ -199,6 +256,44 @@ in another context without any extra changes to its API.
           }
       }
 
+    .. code-tab:: cpp C++
+
+      // Parent
+      get_node<Left>("Left")->target = get_node<Node>("Right/Receiver");
+
+      class Left : public Node {
+        GDCLASS(Left, Node)
```
Comment (AThousandShips) [inline tutorials/best_practices/scene_organization.rst:265]: Same here

File: tutorials/best_practices/scene_organization.rst [hunk 7]
```diff
@@ -69,6 +69,19 @@ initialize it:
        // Child
        EmitSignal("SignalName"); // Triggers parent-defined behavior.
 
+     .. code-tab:: cpp C++
+
+       // Parent
+       Node *node = get_node<Node>("Child");
+       if (node != nullptr) {
+          // Note that get_node may return a nullptr, which would make calling the connect method crash the engine if "Child" does not exist!
```
Comment (AThousandShips) [inline tutorials/best_practices/scene_organization.rst:77]: Same here

File: tutorials/best_practices/scene_organization.rst [hunk 8]
```diff
@@ -88,6 +101,17 @@ initialize it:
        // Child
        Call(MethodName); // Call parent-defined method (which child must own).
 
+     .. code-tab:: cpp C++
+
+       // Parent
+       Node *node = get_node<Node>("Child");
+       if (node != nullptr) {
+          node->set("method_name", "do");
```
Comment (AThousandShips) [inline tutorials/best_practices/scene_organization.rst:109]:
```suggestion
           node->set("method_name", "do");
```

File: tutorials/best_practices/scene_organization.rst [hunk 9]
```diff
@@ -127,6 +162,17 @@ initialize it:
        // Child
        GD.Print(Target); // Use parent-defined node.
 
+     .. code-tab:: cpp C++
+
+       // Parent
+       Node *node = get_node<Node>("Child");
+       if (node != nullptr) {
+          node->set("target", this);
```
Comment (AThousandShips) [inline tutorials/best_practices/scene_organization.rst:170]:
```suggestion
           node->set("target", this);
```

File: tutorials/best_practices/scene_organization.rst [hunk 10]
```diff
@@ -146,6 +192,17 @@ initialize it:
        // Child
        GetNode(TargetPath); // Use parent-defined NodePath.
 
+     .. code-tab:: cpp C++
+
+       // Parent
+       Node *node = get_node<Node>("Child");
+       if (node != nullptr) {
+          node->set("target_path", NodePath(".."));
```
Comment (AThousandShips) [inline tutorials/best_practices/scene_organization.rst:200]:
```suggestion
           node->set("target_path", NodePath(".."));
```

File: tutorials/best_practices/scene_organization.rst [hunk 11]
```diff
@@ -108,6 +132,17 @@ initialize it:
        // Child
        FuncProperty.Call(); // Call parent-defined method (can come from anywhere).
 
+     .. code-tab:: cpp C++
+
+       // Parent
+       Node *node = get_node<Node>("Child");
+       if (node != nullptr) {
+          node->set("func_property", Callable(&ObjectWithMethod::method_on_the_object));
```
Comment (AThousandShips) [inline tutorials/best_practices/scene_organization.rst:140]:
```suggestion
           node->set("func_property", callable_mp(object_with_method, &ObjectWithMethod::method_on_the_object));
```

File: tutorials/best_practices/logic_preferences.rst [hunk 12]
```diff
@@ -100,6 +100,21 @@ either? Let's see an example:
         }
     }
 
+  .. code-tab:: cpp C++
+
+    using namespace godot;
+
+    class MyBuildings : public Node {
```
Comment (AThousandShips) [inline tutorials/best_practices/logic_preferences.rst:107]:
```suggestion
    class MyBuildings : public Node {
        GDCLASS(MyBuildings, Node)

```

File: tutorials/best_practices/godot_notifications.rst [hunk 13]
```diff
@@ -309,3 +379,36 @@ nodes that one might create at runtime.
             GD.Print("I'm reacting to my parent's interaction!");
         }
     }
+
+  .. code-tab:: cpp C++
+
+    using namespace godot;
+
+    class MyNode : public Node {
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:387]:
```suggestion
    class MyNode : public Node {
        GDCLASS(MyNode, Node)

```

File: tutorials/best_practices/godot_notifications.rst [hunk 14]
```diff
@@ -160,6 +181,27 @@ delta time methods as needed.
 
     }
 
+  .. code-tab:: cpp C++
+
+    using namespace godot;
+
+    class MyNode : public Node {
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:188]:
```suggestion
    class MyNode : public Node {
        GDCLASS(MyNode, Node)

```

File: tutorials/best_practices/godot_notifications.rst [hunk 15]
```diff
@@ -102,6 +102,27 @@ implementing a Timer-timeout loop is another option.
         }
     }
 
+ .. code-tab:: cpp C++
+
+    using namespace godot;
+
+    class MyNode : public Node {
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:109]:
```suggestion
    class MyNode : public Node {
        GDCLASS(MyNode, Node)

```

File: tutorials/best_practices/scene_organization.rst [hunk 16]
```diff
@@ -199,6 +256,44 @@ in another context without any extra changes to its API.
...
+
+      public:
+        Node *target = nullptr;
+
+        Left() {}
+
+        void execute() {
+            // Do something with 'target'.
+        }
+
```
Comment (AThousandShips) [inline tutorials/best_practices/scene_organization.rst:278]:
```suggestion
```

File: tutorials/best_practices/scene_organization.rst [hunk 17]
```diff
@@ -199,6 +256,44 @@ in another context without any extra changes to its API.
...
+        static void _bind_methods() {}
+
+      public:
+        Node *receiver = nullptr;
+
+        Right() {
+          receiver = memnew(Node);
+          add_child(receiver);
+        }
+
```
Comment (AThousandShips) [inline tutorials/best_practices/scene_organization.rst:294]:
```suggestion
```

File: tutorials/best_practices/scene_organization.rst [hunk 18]
```diff
@@ -69,6 +69,19 @@ initialize it:
...
        EmitSignal("SignalName"); // Triggers parent-defined behavior.
 
+     .. code-tab:: cpp C++
+
+       // Parent
+       Node *node = get_node<Node>("Child");
+       if (node != nullptr) {
+          // Note that get_node may return a nullptr, which would make calling the connect method crash the engine if "Child" does not exist!
+          // So unless you are 1000% sure get_node will never return a nullptr, it's a good idea to always do a nullptr check.
+          node->connect("signal_name", Callable(this, "method_on_the_object"));
```
Comment (AThousandShips) [inline tutorials/best_practices/scene_organization.rst:79]:
```suggestion
          node->connect("signal_name", callable_mp(this, &ObjectWithMethod::method_on_the_object));
```

File: tutorials/best_practices/godot_notifications.rst [hunk 19]
```diff
@@ -309,3 +379,36 @@ nodes that one might create at runtime.
...
+      void connection_check() {
+          return parent_cache.has_user_signal("interacted_with");
+      }
+
+      void _notification(int p_what) {
+          switch (p_what) {
+              case NOTIFICATION_PARENTED:
+                  parent_cache = get_parent();
+                  if (connection_check()) {
+                      parent_cache.connect("interacted_with", Callable(this, "on_parent_interacted_with"));
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:404]:
```suggestion
                      parent_cache.connect("interacted_with", callable_mp(this, &MyNode::on_parent_interacted_with));
```

File: tutorials/best_practices/godot_notifications.rst [hunk 20]
```diff
@@ -309,3 +379,36 @@ nodes that one might create at runtime.
...
+                  }
+                  break;
+              case NOTIFICATION_UNPARENTED:
+                  if (connection_check()) {
+                      parent_cache.disconnect("interacted_with", Callable(this, "on_parent_interacted_with"));
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:409]:
```suggestion
                      parent_cache.disconnect("interacted_with", callable_mp(this, &MyNode::on_parent_interacted_with));
```

**Review by AThousandShips — CHANGES_REQUESTED** [review]

File: tutorials/best_practices/godot_notifications.rst [hunk 21]
```diff
@@ -160,6 +181,27 @@ delta time methods as needed.
...
+    public:
+        // Called every frame, even when the engine detects no input.
+        void _process(double p_delta) {
+            if (Input::get_singleton->is_action_just_pressed("ui_select"))
+                UtilityFunctions::print(p_delta);
+        }
+
+        // Called during every input event. Equally true for _input().
+        void _unhandled_input(const Ref<InputEvent> &event) {
+            const InputEventKey *key_event = Object::cast_to<const InputEventKey>(*event);
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:198]:
```suggestion
            Ref<InputEventKey> key_event = event;
```

File: tutorials/best_practices/godot_notifications.rst [hunk 22]
```diff
@@ -160,6 +181,27 @@ delta time methods as needed.
...
+            if (key_event != nullptr && Input::get_singleton->is_action_just_pressed("ui_accept")) {
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:199]:
```suggestion
            if (key_event.is_valid() && Input::get_singleton->is_action_just_pressed("ui_accept")) {
```

Comment (dementive) [reply tutorials/best_practices/godot_notifications.rst:199]: If this is the correct way to do this Example::_input in example.cpp in the godot-cpp repository should probably also be changed, that is where I got the cast_to thing from to begin with.

**Comment by dementive** [comment]
I think this should be all good to merge now on my end unless someone finds more problems with it in review. I was planning to do `godot_interfaces.rst` but lost motivation to chug through the example snippets by myself.

**Comment by AThousandShips** [comment]
Will go over this again today

File: tutorials/best_practices/godot_notifications.rst [hunk 23]
```diff
@@ -102,6 +102,28 @@ implementing a Timer-timeout loop is another option.
         }
     }
 
+ .. code-tab:: cpp C++
+
+    using namespace godot;
+
+    class MyNode : public Node {
+        GDCLASS(MyNode, Node)
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:110]:
```suggestion
        GDCLASS(MyNode, Node)

```
Space here, to conform with formatting

File: tutorials/best_practices/godot_notifications.rst [hunk 24]
```diff
@@ -102,6 +102,28 @@ implementing a Timer-timeout loop is another option.
...
+        void _ready() override {
+            Timer *timer = memnew(Timer());
+            timer->set_autostart(true);
+            timer->set_wait_time(0.5);
+            add_child(timer);
+            timer->connect("timeout", Callable(this, "run"));
+        }
+
+        void run() {
+            UtilityFunctions::print("This block runs every 0.5 seconds");
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:123]:
```suggestion
            UtilityFunctions::print("This block runs every 0.5 seconds.");
```
Complete sentences

File: tutorials/best_practices/godot_notifications.rst [hunk 25]
```diff
@@ -102,6 +102,28 @@ implementing a Timer-timeout loop is another option.
...
+        GDCLASS(MyNode, Node)
+    public:
+        // Allows for recurring operations that don't trigger script logic
+        // every frame (or even every fixed frame).
+        void _ready() override {
+            Timer *timer = memnew(Timer());
+            timer->set_autostart(true);
+            timer->set_wait_time(0.5);
+            add_child(timer);
+            timer->connect("timeout", Callable(this, "run"));
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:119]:
```suggestion
            timer->connect("timeout", callable_mp(this, &MyNode::run));
```
Preferred syntax

File: tutorials/best_practices/godot_notifications.rst [hunk 26]
```diff
@@ -102,6 +102,28 @@ implementing a Timer-timeout loop is another option.
...
+ .. code-tab:: cpp C++
+
+    using namespace godot;
+
+    class MyNode : public Node {
+        GDCLASS(MyNode, Node)
+    public:
+        // Allows for recurring operations that don't trigger script logic
+        // every frame (or even every fixed frame).
+        void _ready() override {
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:114]:
```suggestion
        virtual void _ready() override {
```
Standard formatting

File: tutorials/best_practices/godot_notifications.rst [hunk 27]
```diff
@@ -160,6 +182,29 @@ delta time methods as needed.
 
     }
 
+  .. code-tab:: cpp C++
+
+    using namespace godot;
+
+    class MyNode : public Node {
+        GDCLASS(MyNode, Node)
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:190]:
```suggestion
        GDCLASS(MyNode, Node)

```

File: tutorials/best_practices/godot_notifications.rst [hunk 28]
```diff
@@ -223,6 +268,34 @@ values will set up according to the following sequence:
...
+        String test = "one";
+
+    protected:
+        static void _bind_methods() {
+            ClassDB::bind_method(D_METHOD("get_test"), &MyNode::get_test);
+            ClassDB::bind_method(D_METHOD("set_test", "p_test"), &MyNode::set_test);
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:282]:
```suggestion
            ClassDB::bind_method(D_METHOD("set_test", "test"), &MyNode::set_test);
```
Binds do not use the prefix

File: tutorials/best_practices/godot_notifications.rst [hunk 29]
```diff
@@ -223,6 +268,34 @@ values will set up according to the following sequence:
         // the setter, changing _test's value from "two!" to "three!".
     }
 
+  .. code-tab:: cpp C++
+
+    using namespace godot;
+
+    class MyNode : public Node {
+        GDCLASS(MyNode, Node)
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:276]:
```suggestion
        GDCLASS(MyNode, Node)

```

File: tutorials/best_practices/godot_notifications.rst [hunk 30]
```diff
@@ -309,3 +382,37 @@ nodes that one might create at runtime.
             GD.Print("I'm reacting to my parent's interaction!");
         }
     }
+
+  .. code-tab:: cpp C++
+
+    using namespace godot;
+
+    class MyNode : public Node {
+        GDCLASS(MyNode, Node)
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:391]:
```suggestion
        GDCLASS(MyNode, Node)

```

File: tutorials/best_practices/logic_preferences.rst [hunk 31]
```diff
@@ -100,6 +100,22 @@ either? Let's see an example:
...
+        GDCLASS(MyBuildings, Node)
```
Comment (AThousandShips) [inline tutorials/best_practices/logic_preferences.rst:108]:
```suggestion
        GDCLASS(MyBuildings, Node)

```

File: tutorials/best_practices/godot_notifications.rst [hunk 32]
```diff
@@ -102,6 +102,28 @@ implementing a Timer-timeout loop is another option.
...
+    public:
+        // Allows for recurring operations that don't trigger script logic
+        // every frame (or even every fixed frame).
+        void _ready() override {
+            Timer *timer = memnew(Timer());
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:115]:
```suggestion
            Timer *timer = memnew(Timer);
```

File: tutorials/best_practices/godot_notifications.rst [hunk 33]
```diff
@@ -309,3 +382,37 @@ nodes that one might create at runtime.
...
         }
     }
+
+  .. code-tab:: cpp C++
+
+    using namespace godot;
+
+    class MyNode : public Node {
+        GDCLASS(MyNode, Node)
+        Node parent_cache;
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:392]:
```suggestion
        Node *parent_cache = nullptr;
```
`Node` is a pointer type

File: tutorials/best_practices/godot_notifications.rst [hunk 34]
```diff
@@ -309,3 +382,37 @@ nodes that one might create at runtime.
...
+
+        void on_parent_interacted_with() {
+            UtilityFunctions::print("I'm reacting to my parent's interaction!");
+        }
+
+    public:
+        void connection_check() {
+            return parent_cache.has_user_signal("interacted_with");
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:400]:
```suggestion
            return parent_cache->has_user_signal("interacted_with");
```

File: tutorials/best_practices/godot_notifications.rst [hunk 35]
```diff
@@ -309,3 +382,37 @@ nodes that one might create at runtime.
...
+        }
+
+        void _notification(int p_what) {
+            switch (p_what) {
+                case NOTIFICATION_PARENTED:
+                    parent_cache = get_parent();
+                    if (connection_check()) {
+                        parent_cache.connect("interacted_with", callable_mp(this, &MyNode::on_parent_interacted_with));
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:408]:
```suggestion
                        parent_cache->connect("interacted_with", callable_mp(this, &MyNode::on_parent_interacted_with));
```

File: tutorials/best_practices/godot_notifications.rst [hunk 36]
```diff
@@ -309,3 +382,37 @@ nodes that one might create at runtime.
...
+                    }
+                    break;
+                case NOTIFICATION_UNPARENTED:
+                    if (connection_check()) {
+                        parent_cache.disconnect("interacted_with", callable_mp(this, &MyNode::on_parent_interacted_with));
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:413]:
```suggestion
                        parent_cache->disconnect("interacted_with", callable_mp(this, &MyNode::on_parent_interacted_with));
```

File: tutorials/best_practices/godot_notifications.rst [hunk 37]
```diff
@@ -160,6 +182,29 @@ delta time methods as needed.
...
+    public:
+        // Called every frame, even when the engine detects no input.
+        void _process(double p_delta) {
+            if (Input::get_singleton->is_action_just_pressed("ui_select")) {
+                UtilityFunctions::print(p_delta);
+            }
+        }
+
+        // Called during every input event. Equally true for _input().
+        void _unhandled_input(const Ref<InputEvent> &p_event) {
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:200]:
```suggestion
        virtual void _unhandled_input(const Ref<InputEvent> &p_event) override {
```

File: tutorials/best_practices/godot_notifications.rst [hunk 38]
```diff
@@ -160,6 +182,29 @@ delta time methods as needed.
...
 
+  .. code-tab:: cpp C++
+
+    using namespace godot;
+
+    class MyNode : public Node {
+        GDCLASS(MyNode, Node)
+    public:
+        // Called every frame, even when the engine detects no input.
+        void _process(double p_delta) {
```
Comment (AThousandShips) [inline tutorials/best_practices/godot_notifications.rst:193]:
```suggestion
        virtual void _process(double p_delta) override {
```

File: tutorials/best_practices/logic_preferences.rst [hunk 39]
```diff
@@ -100,6 +100,22 @@ either? Let's see an example:
...
+    public:
+        const Ref<PackedScene> building = ResourceLoader::get_singleton()->load("res://building.tscn");
+        Ref<PackedScene> a_building;
+
+        void _ready() override {
```
Comment (AThousandShips) [inline tutorials/best_practices/logic_preferences.rst:113]:
```suggestion
        virtual void _ready() override {
```

**Review by AThousandShips — COMMENTED** [review]
Can't currently speak to the correctness, but this should be the final pass of formatting

**Comment by dsnopek** [comment]
Thanks, these changes seem good to me!

However, this'll need to be squashed to a single commit in order to be merged. See https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase

**Comment by dementive** [comment]
I rebased/squashed all the commits from this PR and did the same for https://github.com/godotengine/godot-docs/pull/9821, both PRs should both be good to go now.

**Review by mhilbrunner — APPROVED** [review]

**Comment by mhilbrunner** [comment]
Merged. Thanks and congrats on your first merged contribution! Amazing work :)

**Pull request merged**